File: cs/temp-launcher/app_pojavlauncher/src/main/java/net/kdt/pojavlaunch/yggdrasil/SkinManager.java
package net.kdt.pojavlaunch.yggdrasil;

import android.util.Log;
import java.io.File;
import java.io.FileInputStream;
import java.io.ByteArrayOutputStream;

public class SkinManager {
    private static final String TAG = "SkinManager";

    public interface SkinAnalyzerFacade {
        PlayerSkin prepareSkin(byte[] bytes);
        PlayerCape prepareCape(byte[] bytes);
    }

    private final SkinAnalyzerFacade analyzer;
    private final OfflineYggdrasilServer server = new OfflineYggdrasilServer();
    private int port = 0;

    public SkinManager(SkinAnalyzerFacade analyzer) {
        this.analyzer = analyzer;
    }

    private static byte[] readFileBytes(File file) {
        if (file == null || !file.exists()) return null;
        try (FileInputStream fis = new FileInputStream(file);
             ByteArrayOutputStream bos = new ByteArrayOutputStream()) {
            byte[] buf = new byte[8192];
            int read;
            while ((read = fis.read(buf)) != -1) {
                bos.write(buf, 0, read);
            }
            return bos.toByteArray();
        } catch (Exception e) {
            return null;
        }
    }

    /**
     * Prepares an account using the account's actual UUID.
     * This ensures the UUID registered in the Yggdrasil server matches
     * the UUID that Minecraft uses in its session requests.
     */
    public void prepareAccountWithUuid(
            String username,
            String profileId,
            File skinFile,
            File capeFile,
            SkinModelType modelOverride
    ) throws InvalidSkinException {
        byte[] skinBytes = readFileBytes(skinFile);
        byte[] capeBytes = readFileBytes(capeFile);

        Log.i(TAG, "Preparing account with explicit UUID: " + profileId);
        Log.i(TAG, "Skin bytes: " + (skinBytes != null ? skinBytes.length + " bytes" : "null"));
        Log.i(TAG, "Cape bytes: " + (capeBytes != null ? capeBytes.length + " bytes" : "null"));

        PlayerSkin skin = null;
        if (skinBytes != null) {
            PlayerSkin base = analyzer.prepareSkin(skinBytes);
            if (base == null) {
                throw new InvalidSkinException((skinFile != null ? skinFile.getName() : "Skin file") + " must be 64x64 or 64x32 pixels");
            }
            if (modelOverride != null && modelOverride != base.getModel()) {
                skin = new PlayerSkin(base.getBytes(), base.getHash(), modelOverride);
            } else {
                skin = base;
            }
            Log.i(TAG, "Skin prepared: hash=" + skin.getHash() + " model=" + skin.getModel());
        }

        PlayerCape cape = null;
        if (capeBytes != null) {
            cape = analyzer.prepareCape(capeBytes);
            if (cape != null) {
                Log.i(TAG, "Cape prepared: hash=" + cape.getHash());
            }
        }

        // Use the provided profileId directly — do NOT regenerate it
        server.addCharacter(username, profileId, skin, cape);
        Log.i(TAG, "Character registered in server: " + username + " / " + profileId);
    }

    public PreparedAccount prepareAccount(
            String username,
            File skinFile,
            File capeFile,
            SkinModelType modelOverride
    ) throws InvalidSkinException {
        byte[] skinBytes = readFileBytes(skinFile);
        byte[] capeBytes = readFileBytes(capeFile);

        PlayerSkin skin = null;
        if (skinBytes != null) {
            PlayerSkin base = analyzer.prepareSkin(skinBytes);
            if (base == null) {
                throw new InvalidSkinException((skinFile != null ? skinFile.getName() : "Skin file") + " must be 64x64 or 64x32 pixels");
            }
            if (modelOverride != null && modelOverride != base.getModel()) {
                skin = new PlayerSkin(base.getBytes(), base.getHash(), modelOverride);
            } else {
                skin = base;
            }
        }

        PlayerCape cape = null;
        if (capeBytes != null) {
            cape = analyzer.prepareCape(capeBytes);
        }

        SkinModelType model = (skin != null) ? skin.getModel() : SkinModelType.NONE;
        String profileId = LocalUuidUtils.generateProfileId(username, model);

        server.addCharacter(username, profileId, skin, cape);

        return new PreparedAccount(
                username,
                profileId,
                LocalUuidUtils.toFormattedUuid(profileId),
                model
        );
    }

    public int startServer() {
        port = server.start();
        return port;
    }

    public void stopServer() {
        server.stop();
    }

    public String getAuthlibUrl() {
        return "http://127.0.0.1:" + port;
    }
    
    public int getPort() {
        return port;
    }

    public static final SkinAnalyzerFacade androidSkinAnalyzerFacade = new SkinAnalyzerFacade() {
        @Override
        public PlayerSkin prepareSkin(byte[] bytes) {
            return SkinAnalyzer.prepareSkin(bytes);
        }

        @Override
        public PlayerCape prepareCape(byte[] bytes) {
            return SkinAnalyzer.prepareCape(bytes);
        }
    };
}

File: temp-launcher/app_pojavlauncher/src/main/java/net/kdt/pojavlaunch/skin/SkinManager.kt
package net.kdt.pojavlaunch.skin

import net.kdt.pojavlaunch.skin.LocalUuidUtils.toFormattedUuid
import java.io.File

/**
 * Facade over skin validation, UUID generation, and the Yggdrasil server.
 */
class SkinManager(private val analyzer: SkinAnalyzerFacade) {

    /**
     * Platform-agnostic interface. Implement using [AndroidSkinAnalyzer]
     */
    interface SkinAnalyzerFacade {
        /** Return null if bytes are not a valid Minecraft skin (wrong dimensions). */
        fun prepareSkin(bytes: ByteArray): PlayerSkin?
        fun prepareCape(bytes: ByteArray): PlayerCape
    }

    private val server = OfflineYggdrasilServer()
    private var port: Int = 0

    /**
     * Validate, hash, and register an offline account.
     *
     * @param modelOverride  Force STEVE or ALEX instead of auto-detecting.
     *                       Pass null (default) to auto-detect from the skin image.
     * @throws InvalidSkinException if the skin file exists but is not 64×64 or 64×32.
     */
    @Throws(InvalidSkinException::class)
    fun prepareAccount(
        username: String,
        skinFile: File? = null,
        capeFile: File? = null,
        modelOverride: SkinModelType? = null
    ): PreparedAccount {
        val skinBytes = skinFile?.takeIf { it.exists() }?.readBytes()
        val capeBytes = capeFile?.takeIf { it.exists() }?.readBytes()

        val skin: PlayerSkin? = skinBytes?.let {
            val base = analyzer.prepareSkin(it)
                ?: throw InvalidSkinException("${skinFile?.name ?: "Skin file"} must be 64×64 or 64×32 pixels")

            if (modelOverride != null && modelOverride != base.model)
                base.copy(model = modelOverride)
            else base
        }

        val cape: PlayerCape? = capeBytes?.let { analyzer.prepareCape(it) }

        val model = skin?.model ?: SkinModelType.NONE
        val profileId = LocalUuidUtils.generateProfileId(username, model)

        server.addCharacter(
            username  = username,
            profileId = profileId,
            skin      = skin,
            cape      = cape
        )

        return PreparedAccount(
            username      = username,
            profileId     = profileId,
            formattedUuid = profileId.toFormattedUuid(),
            skinModel     = model
        )
    }

    /** Start the local server. Call this before launching the game. */
    fun startServer(): Int {
        port = server.start()
        return port
    }

    /** Stop the server once the game process has exited. */
    fun stopServer() = server.stop()

    /** The JVM argument base URL for authlib-injector. */
    val authlibUrl: String get() = "http://127.0.0.1:$port"
}

/** Returned by [SkinManager.prepareAccount]; pass these values to Minecraft's launch args. */
data class PreparedAccount(
    val username: String,
    val profileId: String,
    val formattedUuid: String,
    val skinModel: SkinModelType
)

class InvalidSkinException(message: String) : Exception(message)

val androidSkinAnalyzerFacade = object : SkinManager.SkinAnalyzerFacade {
    override fun prepareSkin(bytes: ByteArray) = AndroidSkinAnalyzer.prepareSkin(bytes)
    override fun prepareCape(bytes: ByteArray) = AndroidSkinAnalyzer.prepareCape(bytes)
}
